pkg/apis/common/v1/validations.go

```diff
@@ -20,7 +21,9 @@ func NoUnknownFields(dest interface{}, meta metav1.ObjectMeta) field.ErrorList {
 	if cfg, ok := meta.Annotations[v1.LastAppliedConfigAnnotation]; ok {
 		d := json.NewDecoder(strings.NewReader(cfg))
 		d.DisallowUnknownFields()
-		if err := d.Decode(&dest); err != nil {
+		// copy the resource to be validated to avoid mutation if the object in the annotation is different
+		destCopy := dest.(runtime.Object).DeepCopyObject()
```

Comment:
How about we change the signature of `NoUnknownFields ` to accept a `runtime.Object` instead of an `interface{}`?
This would avoid the need for casting.